File: packages/agents/src/orchestration/SwarmController.ts
/**
 * @file SwarmController.ts
 * @module packages/agents/src/orchestration/SwarmController
 *
 * WHAT: Advanced Multi-Model Swarm Orchestration system.
 * Coordinates a team of diverse LLMs (Claude, GPT, Gemini, Qwen) on a shared task.
 *
 * WHY: Leverages the unique strengths of different model architectures.
 * Provides a configurable and modular autonomous loop with oversight.
 *
 * HOW:
 * 1. Maintain a 'Swarm' of active models and their current roles.
 * 2. Manage a shared 'Neural Transcript' of the collective conversation.
 * 3. Use a 'Supervisor' model to plan and a 'Critic' model to evaluate completion.
 * 4. Support rotating Implementers, Testers, and Planners.
 */

import { LLMService, SelectedModel } from "@hypercode/ai";
import type { IMCPServer } from "@hypercode/adk";
import { A2AMessageType } from "@hypercode/adk";
import { a2aBroker } from "./A2ABroker.js";

export enum SwarmRole {
    SUPERVISOR = 'supervisor',
    CRITIC = 'critic',
    PLANNER = 'planner',
    IMPLEMENTER = 'implementer',
    TESTER = 'tester'
}

export interface SwarmMember {
    id: string;
    name: string;
    role: SwarmRole;
    provider: string;
    modelId: string;
    status: 'idle' | 'thinking' | 'working';
}

export interface SwarmSessionConfig {
    maxTurns: number;
    completionThreshold: number; // 0.0 to 1.0
    autoRotate: boolean;
}

export class SwarmController {
    private members: Map<string, SwarmMember> = new Map();
    private transcript: string[] = [];
    private activeGoal: string = "";

    constructor(
        private server: IMCPServer,
        private llmService: LLMService
    ) {}

    /**
     * Add a model to the active swarm
     */
    public addMember(member: SwarmMember) {
        this.members.set(member.id, member);
        console.log(`[Swarm] Added member: ${member.name} (${member.role})`);
        
        // Register with A2A broker for communication
        // a2aBroker.registerAgent(member.id, this.createA2AWrapper(member));
    }

    /**
     * Start a swarm collaboration session
     */
    public async startSession(goal: string, config: SwarmSessionConfig) {
        this.activeGoal = goal;
        this.transcript = [`Collective Goal: ${goal}`];
        console.log(`[Swarm] 🧠 Starting session for goal: "${goal}"`);

        let turnCount = 0;
        let isComplete = false;

        while (turnCount < config.maxTurns && !isComplete) {
            turnCount++;
            console.log(`[Swarm] --- Turn ${turnCount} ---`);

            // 1. Planning Turn
            const plan = await this.executeMemberTurn(SwarmRole.PLANNER, "Create the current implementation strategy.");
            this.transcript.push(`PLANNER: ${plan}`);

            // 2. Implementation Turn
            const work = await this.executeMemberTurn(SwarmRole.IMPLEMENTER, `Execute the plan: ${plan}`);
            this.transcript.push(`IMPLEMENTER: ${work}`);

            // 3. Testing Turn
            const testResult = await this.executeMemberTurn(SwarmRole.TESTER, `Verify the work: ${work}`);
            this.transcript.push(`TESTER: ${testResult}`);

            // 4. Evaluation (Critic)
            const evaluation = await this.executeEvaluation(config.completionThreshold);
            this.transcript.push(`CRITIC: ${evaluation.feedback}`);
            
            isComplete = evaluation.isComplete;
            
            if (config.autoRotate) {
                this.rotateRoles();
            }

            // Emit update to dashboard
            await this.broadcastUpdate();
        }

        return {
            success: isComplete,
            turns: turnCount,
            transcript: this.transcript
        };
    }

    private async executeMemberTurn(role: SwarmRole, instruction: string): Promise<string> {
        const member = Array.from(this.members.values()).find(m => m.role === role);
        if (!member) return `[System]: No active member for role ${role}`;

        member.status = 'thinking';
        const prompt = `
            TRANSCRIPT:
            ${this.transcript.join("\n\n")}

            INSTRUCTION for ${member.name} (${member.role.toUpperCase()}):
            ${instruction}
        `;

        try {
            const response = await this.llmService.generateText(
                member.provider,
                member.modelId,
                `You are part of a model swarm. Your role is ${member.role}.`,
                prompt
            );
            member.status = 'idle';
            return response.content;
        } catch (e: any) {
            member.status = 'idle';
            return `ERROR: ${e.message}`;
        }
    }

    private async executeEvaluation(threshold: number): Promise<{ isComplete: boolean, feedback: string }> {
        const critic = Array.from(this.members.values()).find(m => m.role === SwarmRole.CRITIC);
        const modelId = critic?.modelId || "gemini-2.5-flash"; // Fallback to local or cheap model

        const prompt = `
            Evaluate the following swarm transcript against the goal: "${this.activeGoal}"
            
            TRANSCRIPT:
            ${this.transcript.slice(-5).join("\n\n")}

            Is the task complete? If so, start your response with "COMPLETE".
            Otherwise, provide constructive criticism for the next cycle.
        `;

        try {
            const response = await this.llmService.generateText(
                critic?.provider || "google",
                modelId,
                "You are the Swarm Critic.",
                prompt
            );
            const content = response.content.trim();
            return {
                isComplete: content.startsWith("COMPLETE"),
                feedback: content
            };
        } catch (e: any) {
            return { isComplete: false, feedback: "Evaluation failed: " + e.message };
        }
    }

    private rotateRoles() {
        const members = Array.from(this.members.values());
        // Simple rotation logic
        const firstRole = members[0].role;
        for (let i = 0; i < members.length - 1; i++) {
            members[i].role = members[i+1].role;
        }
        members[members.length - 1].role = firstRole;
    }

    private async broadcastUpdate() {
        // Send status to dashboard via event bus or A2A
        await a2aBroker.routeMessage({
            id: `swarm-update-${Date.now()}`,
            timestamp: Date.now(),
            sender: 'SWARM_CONTROLLER',
            type: A2AMessageType.STATE_UPDATE,
            payload: {
                members: Array.from(this.members.values()),
                transcriptCount: this.transcript.length
            }
        });
    }
}

File: go/internal/orchestration/swarm_controller.go
package orchestration

/**
 * @file swarm_controller.go
 * @module go/internal/orchestration
 *
 * WHAT: Go-native implementation of the Multi-Model Swarm Orchestration system.
 * Coordinates a team of LLMs on a shared goal with a shared transcript.
 *
 * WHY: Total Autonomy — The Go sidecar should be capable of managing 
 * complex multi-model workflows independently of the Node control plane.
 */

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"github.com/hypercodehq/hypercode-go/internal/ai"
)

type SwarmRole string

const (
	SwarmRoleSupervisor  SwarmRole = "supervisor"
	SwarmRoleCritic      SwarmRole = "critic"
	SwarmRolePlanner     SwarmRole = "planner"
	SwarmRoleImplementer SwarmRole = "implementer"
	SwarmRoleTester      SwarmRole = "tester"
)

type SwarmMember struct {
	ID       string    `json:"id"`
	Name     string    `json:"name"`
	Role     SwarmRole `json:"role"`
	Provider string    `json:"provider"`
	ModelID  string    `json:"modelId"`
	Status   string    `json:"status"` // "idle", "thinking", "working"
}

type SwarmSessionConfig struct {
	MaxTurns            int     `json:"maxTurns"`
	CompletionThreshold float64 `json:"completionThreshold"`
	AutoRotate          bool    `json:"autoRotate"`
}

type SwarmSessionResult struct {
	Success    bool     `json:"success"`
	Turns      int      `json:"turns"`
	Transcript []string `json:"transcript"`
}

type SwarmController struct {
	mu         sync.RWMutex
	members    map[string]*SwarmMember
	transcript []string
	activeGoal string
	broker     *A2ABroker
}

func NewSwarmController(broker *A2ABroker) *SwarmController {
	return &SwarmController{
		members:    make(map[string]*SwarmMember),
		transcript: make([]string, 0),
		broker:     broker,
	}
}

func (c *SwarmController) AddMember(member SwarmMember) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.members[member.ID] = &member
	fmt.Printf("[Go Swarm] Added member: %s (%s)\n", member.Name, member.Role)
}

func (c *SwarmController) StartSession(ctx context.Context, goal string, cfg SwarmSessionConfig) (*SwarmSessionResult, error) {
	c.mu.Lock()
	c.activeGoal = goal
	c.transcript = []string{fmt.Sprintf("Collective Goal: %s", goal)}
	c.mu.Unlock()

	fmt.Printf("[Go Swarm] Starting session for goal: %s\n", goal)

	turnCount := 0
	isComplete := false

	for turnCount < cfg.MaxTurns && !isComplete {
		turnCount++
		fmt.Printf("[Go Swarm] --- Turn %d ---\n", turnCount)

		// 1. Planning Turn
		plan, err := c.executeMemberTurn(ctx, SwarmRolePlanner, "Create the current implementation strategy.")
		if err != nil {
			return nil, err
		}
		c.addTranscript(fmt.Sprintf("PLANNER: %s", plan))

		// 2. Implementation Turn
		work, err := c.executeMemberTurn(ctx, SwarmRoleImplementer, fmt.Sprintf("Execute the plan: %s", plan))
		if err != nil {
			return nil, err
		}
		c.addTranscript(fmt.Sprintf("IMPLEMENTER: %s", work))

		// 3. Testing Turn
		testResult, err := c.executeMemberTurn(ctx, SwarmRoleTester, fmt.Sprintf("Verify the work: %s", work))
		if err != nil {
			return nil, err
		}
		c.addTranscript(fmt.Sprintf("TESTER: %s", testResult))

		// 4. Evaluation (Critic)
		evaluation, err := c.executeEvaluation(ctx, cfg.CompletionThreshold)
		if err != nil {
			return nil, err
		}
		c.addTranscript(fmt.Sprintf("CRITIC: %s", evaluation.Feedback))
		
		isComplete = evaluation.IsComplete
		
		if cfg.AutoRotate {
			c.rotateRoles()
		}

		c.broadcastUpdate()
	}

	return &SwarmSessionResult{
		Success:    isComplete,
		Turns:      turnCount,
		Transcript: c.getTranscript(),
	}, nil
}

func (c *SwarmController) executeMemberTurn(ctx context.Context, role SwarmRole, instruction string) (string, error) {
	c.mu.RLock()
	var member *SwarmMember
	for _, m := range c.members {
		if m.Role == role {
			member = m
			break
		}
	}
	transcript := c.transcript
	c.mu.RUnlock()

	if member == nil {
		return fmt.Sprintf("[System]: No active member for role %s", role), nil
	}

	member.Status = "thinking"

	systemPrompt := ai.GetSwarmPrompt(string(role))

	prompt := fmt.Sprintf(`
		TRANSCRIPT:
		%s

		INSTRUCTION for %s (%s):
		%s
	`, strings.Join(transcript, "\n\n"), member.Name, strings.ToUpper(string(member.Role)), instruction)

	messages := []ai.Message{
		{Role: "system", Content: systemPrompt},
		{Role: "user", Content: prompt},
	}

	resp, err := ai.AutoRouteWithModel(ctx, member.ModelID, messages)
	member.Status = "idle"
	if err != nil {
		return "", err
	}

	return resp.Content, nil
}

type SwarmEvaluation struct {
	IsComplete bool
	Feedback   string
}

func (c *SwarmController) executeEvaluation(ctx context.Context, threshold float64) (*SwarmEvaluation, error) {
	c.mu.RLock()
	var critic *SwarmMember
	for _, m := range c.members {
		if m.Role == SwarmRoleCritic {
			critic = m
			break
		}
	}
	transcript := c.transcript
	goal := c.activeGoal
	c.mu.RUnlock()

	modelID := "gemini-2.5-flash"
	if critic != nil {
		modelID = critic.ModelID
	}

	prompt := fmt.Sprintf(`
		Evaluate the following swarm transcript against the goal: "%s"
		
		TRANSCRIPT:
		%s

		Is the task complete? If so, start your response with "COMPLETE".
		Otherwise, provide constructive criticism for the next cycle.
	`, goal, strings.Join(transcript[len(transcript)-min(5, len(transcript)):], "\n\n"))

	messages := []ai.Message{
		{Role: "system", Content: ai.GetSwarmPrompt("critic")},
		{Role: "user", Content: prompt},
	}

	resp, err := ai.AutoRouteWithModel(ctx, modelID, messages)
	if err != nil {
		return nil, err
	}

	content := strings.TrimSpace(resp.Content)
	return &SwarmEvaluation{
		IsComplete: strings.HasPrefix(content, "COMPLETE"),
		Feedback:   content,
	}, nil
}

func (c *SwarmController) rotateRoles() {
	c.mu.Lock()
	defer c.mu.Unlock()
	
	ids := make([]string, 0, len(c.members))
	for id := range c.members {
		ids = append(ids, id)
	}
	if len(ids) < 2 {
		return
	}

	firstRole := c.members[ids[0]].Role
	for i := 0; i < len(ids)-1; i++ {
		c.members[ids[i]].Role = c.members[ids[i+1]].Role
	}
	c.members[ids[len(ids)-1]].Role = firstRole
}

func (c *SwarmController) addTranscript(entry string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.transcript = append(c.transcript, entry)
}

func (c *SwarmController) getTranscript() []string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	return append([]string(nil), c.transcript...)
}

func (c *SwarmController) broadcastUpdate() {
	c.mu.RLock()
	members := make([]SwarmMember, 0, len(c.members))
	for _, m := range c.members {
		members = append(members, *m)
	}
	transcriptCount := len(c.transcript)
	c.mu.RUnlock()

	c.broker.RouteMessage(A2AMessage{
		ID:        fmt.Sprintf("swarm-update-%d", nowMillis()),
		Timestamp: nowMillis(),
		Sender:    "SWARM_CONTROLLER",
		Type:      StateUpdate,
		Payload: map[string]interface{}{
			"members":         members,
			"transcriptCount": transcriptCount,
		},
	})
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func nowMillis() int64 {
	return time.Now().UTC().UnixMilli()
}
